client/src/main/java/org/mvndaemon/mvnd/client/DefaultClient.java

```diff
@@ -58,41 +59,50 @@
     private final DaemonParameters parameters;
 
     public static void main(String[] argv) throws Exception {
-        final List<String> args = new ArrayList<>(argv.length);
+        final List<String> args = new ArrayList<>(Arrays.asList(argv));
 
+        // Log file
         Path logFile = null;
-        int i = 0;
-        boolean batchMode = false;
-        while (i < argv.length) {
-            final String arg = argv[i++];
-            if ("-l".equals(arg) || "--log-file".equals(arg)) {
-                if (i < argv.length) {
-                    logFile = Paths.get(argv[i++]);
-                } else {
-                    throw new IllegalArgumentException("-l and --log-file need to be followed by a path");
-                }
+        String sLogFile = Environment.MAVEN_LOG_FILE.removeCommandLineOption(args);
+        if (sLogFile != null) {
+            if (sLogFile.isEmpty()) {
+                throw new IllegalArgumentException("-l and --log-file need to be followed by a path");
+            } else {
+                logFile = Paths.get(sLogFile);
+            }
+        }
 
-            } else if (arg.startsWith("-D")) {
-                /* This needs to be done very early, otherwise various DeamonParameters do not work properly */
-                final int eqPos = arg.indexOf('=');
-                if (eqPos >= 0) {
-                    System.setProperty(arg.substring(2, eqPos), arg.substring(eqPos + 1));
+        // Batch mode
+        boolean batchMode = Environment.MAVEN_BATCH_MODE.hasCommandOption(args);
+
+        // System properties
+        for (Iterator<String> it = args.iterator(); it.hasNext();) {
+            String arg = it.next();
+            String val;
+            if (arg.startsWith("-D")) {
+                val = arg.substring(2);
+            } else if (arg.equals("--define")) {
+                if (it.hasNext()) {
+                    val = it.next();
                 } else {
-                    System.setProperty(arg.substring(2), "");
+                    throw new IllegalArgumentException("Missing argument for option --define");
                 }
-                args.add(arg);
             } else {
-                if (!batchMode && ("-B".equals(arg) || "--batch-mode".equals(arg))) {
-                    batchMode = true;
-                }
-                args.add(arg);
+                continue;
+            }
+            /* This needs to be done very early, otherwise various DaemonParameters do not work properly */
+            final int eqPos = val.indexOf('=');
+            if (eqPos >= 0) {
+                System.setProperty(val.substring(2, eqPos), val.substring(eqPos + 1));
+            } else {
+                System.setProperty(val.substring(2), "");
```

Comment:
I guess this should be moved under the `if (arg.startsWith("-D")) {` branch above to work correctly.

Reply:
I don't think so.  This code is executed for both `-Dxxx=yyy` and `--define xxx=yyy`.
However, there is a duplication of the same code on lines https://github.com/mvndaemon/mvnd/blob/ab48ddd7157a60adf96da144b36897df2fca3d22/client/src/main/java/org/mvndaemon/mvnd/client/DefaultClient.java#L167-L174 which could be removed.

Reply:
> I don't think so. This code is executed for both `-Dxxx=yyy` and `--define xxx=yyy`.

I see, but then, I guess `arg.substring(2...` should rather be `arg.substring(0...`

> However, there is a duplication of the same code on lines
> 
> https://github.com/mvndaemon/mvnd/blob/ab48ddd7157a60adf96da144b36897df2fca3d22/client/src/main/java/org/mvndaemon/mvnd/client/DefaultClient.java#L167-L174
> 
> which could be removed.

Not sure, how exactly you'd like to remove. Extracting to a method could work, but simply removing one of the two will not. The first occurrence is important for native mode and the second is important for the test client.

Reply:
Right, that's the reason why the tests are failing.  I'll move the code in a method.